aea/aea_builder.py

```diff
@@ -0,0 +1,418 @@
+
+    @property
+    def skills(self) -> Dict[ComponentId, Skill]:
+        """Get the skills."""
+        return cast(
+            Dict[ComponentId, Skill],
+            self._all_dependencies_by_type.get(ComponentType.SKILL, {}),
+        )
+
+    @property
+    def contracts(self) -> Dict[ComponentId, Any]:
+        """Get the contracts."""
+        return cast(
+            Dict[ComponentId, Any],
+            self._all_dependencies_by_type.get(ComponentType.CONTRACT, {}),
+        )
+
+    def add_component(self, component: Component) -> None:
+        """Add a component."""
+        self._dependencies[component.component_id] = component
+        self._all_dependencies_by_type.setdefault(component.component_type, {})[
+            component.component_id
+        ] = component
+        for dependency in component.configuration.package_dependencies:
+            self._inverse_dependency_graph.setdefault(dependency, set()).add(
+                component.component_id
+            )
+
+    def remove_component(self, component_id: ComponentId):
+        """
+        Remove a component.
+
+        :return None
+        :raises ValueError: if some component depends on this package.
+        """
+        if component_id not in self.all_dependencies:
+            raise ValueError(
+                "Component {} of type {} not present.".format(
+                    component_id.public_id, component_id.component_type
+                )
+            )
+        dependencies = self._inverse_dependency_graph.get(component_id, set())
+        if len(dependencies) != 0:
+            raise ValueError(
+                "Cannot remove component {} of type {}. Other components depends on it: {}".format(
+                    component_id.public_id, component_id.component_type, dependencies
+                )
+            )
+
+        # remove from the index of all dependencies
+        component = self._dependencies.pop(component_id)
+        # remove from the index of all dependencies grouped by type
+        self._all_dependencies_by_type[component_id.component_type].pop(component_id)
+        if len(self._all_dependencies_by_type[component_id.component_type]) == 0:
+            self._all_dependencies_by_type.pop(component_id.component_type)
+        # update inverse dependency graph
+        for dependency in component.configuration.package_dependencies:
+            self._inverse_dependency_graph[dependency].discard(component_id)
+
+    def check_package_dependencies(
+        self, component_configuration: ComponentConfiguration
+    ) -> bool:
+        """
+        Check that we have all the dependencies needed to the package.
+
+        return: True if all the dependencies are covered, False otherwise.
+        """
+        not_supported_packages = component_configuration.package_dependencies.difference(
+            self.all_dependencies
+        )
+        return len(not_supported_packages) == 0
+
+    @contextmanager
+    def load_dependencies(self):
```

Comment:
only add the highest version for two otherwise identical component ids

Reply:
provide a useful error message

Reply:
> provide a useful error message
We load only the highest version (see the property `dependencies_highest_version`), but no warning message yet (TODO)